DEV-36849; fix: trigger input event before and after replacement

abhijeetnarvekar opened the pull request:
## Make replacements work in Quip Editor

Quip is a state based editor, it triggers changes to state when it encounters events.
Example: When you press a key on keyboard, beforeInput and input events are triggered on the text node, which then bubble up in the DOM

Our current workflow does not trigger input events on text node, the text appears to change on DOM but the state inside editor doesn't update, resulting loss of changes when browser is reloaded.

## Solution:

- Trigger input events simulating real world scenario, which helps the editor to understand that it needs to update its state
- The input events are specifically triggered on text node as text nodes are the leaf nodes and events can then bubble up in DOM
- The solution implemented is a generic solution, rather than limiting only to Quip
- Config option added to disable input event simulation if not necessary

## Tests:
- Added test on ContentEditable and StateBasedAdapter to verify if simulated events are triggered
- Existing `500+` tests continue to work as expected

loaiabdelhalim commented on src/adapters/AbstractRichtextEditorAdapter.ts:
```diff
@@ -224,12 +224,17 @@ export abstract class AbstractRichtextEditorAdapter implements AdapterInterface
     const replacement = alignedMatches.map((m) => m.originalMatch.replacement).join('');
+
+    // Capturing parent node is necessary, replacement makes the textnode orphan.
+    const parentNode = this.getEditorDocument().getSelection()?.focusNode?.parentNode;
```
What happens if the parentNode is undefined?

abhijeetnarvekar replied:
@loaiabdelhalim It will not trigger the input event and log a warning message to the browser console. https://github.com/acrolinx/sidebar-sdk-js/pull/91/files#diff-5acab41990585fa68ae01ff7b2e2915628d19237840c3061b31fcadc1c0f09b8R49

abhijeetnarvekar replied:
Fixed. Input simulation will be triggered only if parent node is defined

abhijeetnarvekar commented:
@pcdeshmukh 

Addressed the following:

- check parent node is null before triggering event - @loaiabdelhalim suggestion
- renamed fakeInput to simulateInput
- simulated event in pairs wherever used
- async state based adapter test check for both input events

pcdeshmukh commented on src/adapters/AsyncContentEditableAdapter.ts:
```diff
@@ -64,12 +64,12 @@ export class AsyncContentEditableAdapter extends AbstractRichtextEditorAdapter i
     assertElementIsDisplayed(this.getEditorElement());
     const [alignedMatches] = this.selectMatches(checkId, matchesWithReplacement);
     const replacement = alignedMatches.map((m) => m.originalMatch.replacement).join('');
+    simulateInputEvent(this.getEditorElement(), 'beforeinput');
```
Added config check this.config.disableInputEventSimulation

abhijeetnarvekar replied:
Fixed

pcdeshmukh commented on src/adapters/InputAdapter.ts:
```diff
@@ -148,11 +148,12 @@ export class InputAdapter implements AdapterInterface {
   replaceRanges(checkId: string, matchesWithReplacement: MatchWithReplacement[]) {
     const alignedMatches = this.selectMatches(checkId, matchesWithReplacement);
     this.scrollAndSelect(alignedMatches);
+    simulateInputEvent(this.element, 'beforeinput');
```
Add config check this.config.disableInputEventSimulation
